File: Problemset/0946-Validate Stack Sequences/Solution.rb
# @param {Integer[]} pushed
# @param {Integer[]} popped
# @return {Boolean}
def validate_stack_sequences(pushed, popped)
    stack = []
    i = 0

    for val in pushed
        stack.push(val)
        while i < popped.length and stack.last == popped[i]
            stack.pop
            i += 1
        end
    end

    return stack.empty?
end

File: Problemset/0946-Validate Stack Sequences/Solution.rs
impl Solution {
    pub fn validate_stack_sequences(pushed: Vec<i32>, popped: Vec<i32>) -> bool {
        let mut stack = vec![];
        let mut i = 0;

        for val in pushed {
            stack.push(val);
            while i < popped.len() && *stack.last().unwrap_or(&-1) == popped[i] {
                stack.pop();
                i += 1;
            }
        }

        stack.is_empty()
    }
}
